src/core/inventory.ts

```diff
@@ -0,0 +1,371 @@
+  }
+
+  // Calculate percentage distribution for each asset on each chain
+  for (const [chainIdStr, chainConfig] of Object.entries(chains)) {
+    const chainId = parseInt(chainIdStr)
+
+    if (!chainConfig.assets) continue
+
+    for (const [symbol, assetConfig] of Object.entries(chainConfig.assets)) {
+      const weight = getAssetWeight(chainId, symbol)
+      const totalWeight = assetTotalWeights[symbol]
+
+      if (totalWeight > 0) {
+        // Calculate the target percentage for this asset on this chain
+        targetDistribution[chainId][symbol] = weight / totalWeight
+      }
+    }
+  }
+
+  return targetDistribution
+}
+
+// Determine the optimal destination chain for an asset
+export function findUnderfundedChain(
+  asset: string,
+  sourceChainId: number,
+  currentBalances: PortfolioBalance,
+): number | null {
+  // Get target distribution
+  const targetDistribution = calculateTargetDistribution()
+
+  // Calculate actual distribution percentages
+  const actualDistribution: Record<number, Record<string, number>> = {}
+
+  // Calculate total value for this specific asset
+  let totalValue = 0
+  for (const chain of currentBalances.chains) {
+    if (!actualDistribution[chain.chainId]) {
+      actualDistribution[chain.chainId] = {}
+    }
+
+    // Find the asset in this chain's assets (should be at most one entry now)
+    const assetBalance = chain.assets.find((a) => a.symbol === asset)
+    if (assetBalance) {
+      totalValue += assetBalance.valueUSD
+    }
+  }
+
+  // No need to calculate for multiple assets anymore
+  // Calculate actual distribution percentages for this asset
+  for (const chain of currentBalances.chains) {
+    const assetBalance = chain.assets.find((a) => a.symbol === asset)
+
+    if (assetBalance && totalValue > 0) {
+      actualDistribution[chain.chainId][asset] =
+        assetBalance.valueUSD / totalValue
+    } else {
+      actualDistribution[chain.chainId][asset] = 0
+    }
+  }
+
+  // Find the chain with the biggest deficit compared to target
+  let biggestDeficit = -1
+  let optimalChainId: number | null = null
+
+  for (const [chainIdStr, chainTargets] of Object.entries(targetDistribution)) {
+    const chainId = parseInt(chainIdStr)
+
+    // Skip source chain
+    if (chainId === sourceChainId) continue
+
+    // Check if this chain supports this asset
+    if (!(asset in chainTargets)) continue
+
+    const targetPercentage = chainTargets[asset]
+    const actualPercentage = actualDistribution[chainId]?.[asset] || 0
+
+    const deficit = targetPercentage - actualPercentage
+
+    if (deficit > biggestDeficit) {
+      biggestDeficit = deficit
+      optimalChainId = chainId
+    }
+  }
+
+  return optimalChainId
+}
+
+// Get optimal destination chain based on inventory distribution
+// This function uses the depositEvent to determine which asset we're handling
+// and finds the most underfunded chain for that specific asset
+export async function getOptimalDestinationChain(
```

Comment:
how much latency does this add before filling? currently, our filltimes are around 1.5 to 3s so adding a bunch of rpc calls will likely push us out of this range right?

Reply:
you're right. I made a note here about caching. Tbh / IMO we should make a class to manage balances here, and if it handles the outgoing transactions appropriately, we should be able to use expected balance diffs/changes on-chain. We could attach a semaphore/thread lock to it and have it sync in the bg as a macrotask possibly? Open to your thoughts here

Reply:
yeah I think syncing in bg makes the most sense here